sdb/commands/linux/internal/slub_helpers.py

```diff
@@ -122,3 +124,173 @@ def util(cache: drgn.Object) -> int:
+        children.append(child)
+    return children
+
+
+def cache_get_partials(cache: drgn.Object) -> Set[int]:
+    assert cache.type_.type_name() == 'struct kmem_cache *'
+    partials = []
+
+    node = cache.node[0].partial  # assumption nr_node_ids == 0
+    for page in list_for_each_entry("struct page", node.address_of_(), "lru"):
+        partials.append(page.value_())
+
+    if not is_root_cache(cache):
+        return set(partials)
+
+    for child in list_for_each_entry("struct kmem_cache",
+                                     cache.memcg_params.children.address_of_(),
+                                     "memcg_params.children_node"):
+        partials += cache_get_partials(child)
+    return set(partials)
+
+
+def for_each_object_in_cache(cache: drgn.Object) -> drgn.Object:
+    """
+    Goes through each object in the SLUB cache supplied
+    yielding them to the consumer.
+    """
+    pg_slab_flag = 1 << sdb.get_prog().constant('PG_slab').value_()
+
+    cache_children = cache_gather_children(cache)
+    #
+    # We are caching the list of partial slabs here so later we can
+    # decide whether we want to use the freelist to find which entries
+    # are allocated (partial slab case) or just read all the entries
+    # (full slab case). This is not great for live-systems where these
+    # data can be changing very quickly under us but its the best that
+    # we can do given that the Linux kernel doesn't keep track of its
+    # full slabs explicitly except in cases where very heavy-weight
+    # debugging config options are enabled.
+    #
+    cache_partials = cache_get_partials(cache)
+
+    #
+    # For the same reason mentioned above, since the Linux kernel
+    # doesn't keep track of its full slabs by default in a list
+    # as it does for partial slabs, we have to go through all the
+    # pages in memory to find those slabs.
+    #
+    for page in for_each_page(sdb.get_prog()):
+        try:
+            if not page.flags & pg_slab_flag:
+                continue
+        except drgn.FaultError:
+            #
+            # Not 100% sure why this happens but we've verified
+            # in multiple cases that we've been able to detect
+            # all the slabs/pages from all the caches and none
+            # of them fall into these segments that give us
+            # FaultErrors. One hypothesis of why these occur
+            # is because the regions that we've tried to get
+            # rid from where origininally all zeros and they
+            # where subsequently filtered out. Another one is
+            # that these regions are just not mapped to anything.
+            # In any case this scenario hasn't brought up any
+            # issues with the functionality that we are trying
+            # to provide.
+            #
+            continue
+
+        slab_cache = page.slab_cache
+        if slab_cache == sdb.get_typed_null(slab_cache.type_):
```

Comment:
Is there a more elegant way to do this, something like `slab_cache.value_() == 0` ?  The docs say:
> For pointers, this [.value_()] returns the address value of the pointer.